src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/SemanticTokensCache.cs

```diff
@@ -0,0 +1,273 @@
+            {
+                lineToTokensDict = new Dictionary<int, IReadOnlyList<int>>();
+                documentCache.Set(semanticVersion, lineToTokensDict);
+            }
+
+            CacheTokensPerLine(range, tokens, lineToTokensDict);
+            FillInEmptyLineGaps(range, lineToTokensDict);
+        }
+
+        /// <summary>
+        /// Attempts to retrieve the cached tokens for a given document/semanticVersion/range. If the cache does
+        /// not have complete tokens for the range, it's possible to return only tokens associated with a portion
+        /// of the range.
+        /// </summary>
+        /// <param name="uri">The URI associated with the cached tokens.</param>
+        /// <param name="semanticVersion">The semantic version associated with the cached tokens.</param>
+        /// <param name="requestedRange">The requested range for the desired tokens.</param>
+        /// <param name="cachedTokens">If found, contains the cached range and cached tokens.</param>
+        /// <returns>
+        /// True if at least a partial match for the range was found. The 'Range' out var specifies the subset of
+        /// the requested range that was found, and the 'Tokens' out var contains the tokens for that subset range.
+        /// </returns>
+        public bool TryGetCachedTokens(
+            DocumentUri uri,
+            VersionStamp semanticVersion,
+            Range requestedRange,
+            [NotNullWhen(true)] out (Range Range, List<int> Tokens)? cachedTokens)
+        {
+            if (!_cache.TryGetValue(uri, out var documentCache) ||
+                !documentCache.TryGetValue(semanticVersion, out var lineToTokensDict))
+            {
+                // No cached results found
+                cachedTokens = null;
+                return false;
+            }
+
+            // Keep track of where the cached range begins
+            Position? cachedRangeStart = null;
+
+            // Keep track of the length of the cached range
+            var cachedRangeLength = 0;
+
+            var tokens = GetCachedTokens(requestedRange, lineToTokensDict, ref cachedRangeStart, ref cachedRangeLength);
+            if (tokens.Count == 0)
+            {
+                // We couldn't find any tokens associated with the passed-in range
+                cachedTokens = null;
+                return false;
+            }
+
+            Assumes.NotNull(cachedRangeStart);
+
+            // We can potentially return a partial range match if we can't find a full range match.
+            var endPosition = new Position { Line = cachedRangeStart.Line + cachedRangeLength, Character = 0 };
+            var range = new Range { Start = cachedRangeStart, End = endPosition };
+
+            cachedTokens = (range, tokens);
+            return true;
+        }
+
+        private void CacheTokensPerLine(Range range, int[] tokens, Dictionary<int, IReadOnlyList<int>> lineToTokensDict)
```

Comment:
this could use docs.  specifically, is 'tokens' supposed to reprent the entire document?